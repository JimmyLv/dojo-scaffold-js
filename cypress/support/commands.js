// ***********************************************
// This example commands.js shows you how to
// create various custom commands and overwrite
// existing commands.
//
// For more comprehensive examples of custom
// commands please read more here:
// https://on.cypress.io/custom-commands
// ***********************************************
//
//
// -- This is a parent command --
// Cypress.Commands.add("login", (email, password) => { ... })
//
//
// -- This is a child command --
// Cypress.Commands.add("drag", { prevSubject: 'element'}, (subject, options) => { ... })
//
//
// -- This is a dual command --
// Cypress.Commands.add("dismiss", { prevSubject: 'optional'}, (subject, options) => { ... })
//
//
// -- This is will overwrite an existing command --
// Cypress.Commands.overwrite("visit", (originalFn, url, options) => { ... })
import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'

// this is a handy function that I normally make available for all my tests
// that deal with connected components.
// you can provide initialState or the entire store that the ui is rendered with
const mockStore = configureMockStore()

Cypress.Commands.add('mountWithRedux', (
  ui,
  data = {},
) => {
  const { initialState, store = mockStore(initialState) } = data
  cy.spy(store, 'dispatch').as('dispatch')
  cy.mount(<Provider store={store}>{ui}</Provider>)
})
