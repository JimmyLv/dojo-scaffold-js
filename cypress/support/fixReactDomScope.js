import ReactDom from 'react-dom';
// see https://github.com/bahmutov/cypress-react-unit-test/issues/51
export default function fixReactDomScope(win) {
  if (win.ReactDOM !== ReactDom) {
    win.ReactDOM = ReactDom;
  }
}
