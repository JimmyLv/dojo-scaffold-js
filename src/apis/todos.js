export function query(pagination) {
  return fetch(`/api/todos`).then(response => response.json())
}
