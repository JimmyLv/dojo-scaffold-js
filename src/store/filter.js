export const NAME = 'filter'
export const types = {
  FILTER: `${NAME}/SET_FILTER`,
}

export default {
  namespace: NAME,
  actions: {
    setFilterStatus: status => ({
      type: types.FILTER,
      payload: { status },
    }),
  },
  reducer(state = 'All', action) {
    const { type, payload } = action
    switch (type) {
      case types.FILTER:
        return payload.status
      default:
        return state
    }
  },
  selectors: {},
}
