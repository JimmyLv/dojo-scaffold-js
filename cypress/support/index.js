// ***********************************************************
// This example support/index.js is processed and
// loaded automatically before your test files.
//
// This is a great place to put global configuration and
// behavior that modifies Cypress.
//
// You can change the location of this file or turn off
// automatically serving support files with the
// 'supportFile' configuration option.
//
// You can read more here:
// https://on.cypress.io/configuration
// ***********************************************************

import 'cypress-jest-adapter'
import 'cypress-plugin-snapshots/commands'

import 'cypress-react-unit-test'
import * as ReactDOM from 'react-dom'
import './commands'

// Import commands.js using ES2015 syntax:
// import 'cypress-skip-and-only-ui/support'
// import '@cypress/code-coverage/support'

// Alternatively you can use CommonJS syntax:
// require('./commands')

// Fixes incorrect React version
// https://github.com/bahmutov/cypress-react-unit-test/issues/51
Cypress.on('window:load', win => {
  win.ReactDOM = window.ReactDOM || win.ReactDOM
})

export function fixReactDomScope(win) {
  if (win.ReactDOM !== ReactDOM) {
    win.ReactDOM = ReactDOM
  }
}

export const fixCypressSpec = (filename, window) => () => {
  fixReactDomScope(window)
  const path = require('path')
  const relative = filename.substr(1) // removes leading "/"
  const projectRoot = Cypress.config('projectRoot')
  const absolute = path.join(projectRoot, relative)
  Cypress.spec = {
    absolute,
    name: path.basename(filename),
    relative,
  }
}
