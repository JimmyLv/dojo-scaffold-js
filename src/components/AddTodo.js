import React, { useState } from 'react'
import { connect } from 'react-redux'
import { actions } from '../store/todos'

export function AddTodo({ dispatch }) {
  const [value, setValue] = useState('')

  const handleSubmit = e => {
    dispatch(actions.addTodo(value))
    setValue('')
    e.preventDefault()
  }

  return <form onSubmit={handleSubmit}>
    <input
      type="text"
      placeholder={'Do that thing...'}
      value={value}
      onChange={e => setValue(e.target.value)}
    />
    <button type="submit">+</button>
  </form>
}

export default connect()(AddTodo)
