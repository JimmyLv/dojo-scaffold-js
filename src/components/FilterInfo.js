import React from 'react'
import { connect } from 'react-redux'
import { actions } from '../store/filter'

export function FilterInfo({ filters = [], dispatch }) {
  return (
    <div>
      {filters.map(status => (
        <span key={status}>
          <a
            href={`#${status}`}
            onClick={() => dispatch(actions.setFilterStatus(status))}
          >
            {status}
          </a>{' '}
        </span>
      ))}
    </div>
  )
}

export default connect()(FilterInfo)
