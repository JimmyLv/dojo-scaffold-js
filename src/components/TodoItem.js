import React from 'react'
import { useDispatch } from 'react-redux'
import styled from 'styled-components'
import { actions } from '../store/todo'

const StyledItem = styled.li.attrs(props => ({
  className: props.completed ? 'done' : undefined,
}))`
  color: ${props => (props.completed ? '#d9d9d9' : '#000')};
  text-decoration: ${props => (props.completed ? 'line-through' : undefined)};
`

function TodoItem({ id, text, completed }) {
  const dispatch = useDispatch()
  return (
    <StyledItem completed={completed}>
      <span onClick={() => dispatch(actions.toggle(id))}>{`${text} `}</span>
      <span className="delete" onClick={() => dispatch(actions.remove(id))}>
        x
      </span>
    </StyledItem>
  )
}

export default TodoItem
