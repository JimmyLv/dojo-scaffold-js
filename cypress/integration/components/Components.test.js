import React from 'react'
import AddTodo from '../../../src/components/AddTodo'
import FilterInfo from '../../../src/components/FilterInfo'
import TodoItem from '../../../src/components/TodoItem'
import * as filter from '../../../src/store/filter'
import * as todos from '../../../src/store/todos'
import { fixCypressSpec } from '../../support'
import fixReactDomScope from '../../support/fixReactDomScope'

beforeEach(fixCypressSpec(__filename, window))

describe('TodoItem Component', () => {

  it('shows an item', () => {
    // given
    const props = {
      id: 1,
      text: 'test item',
      completed: false,
    }

    // when
    cy.mountWithRedux(<TodoItem {...props} />)

    // then
    cy.contains('test item')
      .toMatchImageSnapshot()
  })

  it('marks done items', () => {
    // given
    const props = {
      id: 1,
      text: 'test item',
      completed: true,
    }

    // when
    cy.mountWithRedux(<TodoItem {...props} />)

    // then
    cy.contains('test item').parent().should('have.class', 'done')
  })

  it('toggle completed on click', () => {
    // given
    const props = {
      id: 1,
      text: 'test item',
      completed: false,
    }
    // when
    cy.mountWithRedux(<TodoItem {...props} />)

    // then
    cy.contains('test item').click()
    // just verify the dispatched action
    cy.get('@dispatch').should('be.calledWith', {
      type: todos.types.TOGGLE,
      payload: { id: 1 },
    })
  })

  it('remove item on click', () => {
    // given
    const props = {
      id: 1,
      text: 'test item',
      completed: true,
    }

    // when
    cy.mountWithRedux(<TodoItem {...props} />)

    // then
    cy.contains('test item').siblings('.delete').click()
    // just verify the dispatched action
    cy.get('@dispatch').should('be.calledWith', {
      type: todos.types.REMOVE,
      payload: { id: 1 },
    })
  })
})

describe('AddTodo Component', () => {
  // beforeEach(() => { fixReactDomScope(window) })

  it('add new todo item', () => {
    // given
    const TODO_ITEM = 'new todo.'
    // when
    cy.mountWithRedux(<AddTodo />)

    // then
    // cy.get(AddTodo).toMatchSnapshot()
    cy.get('[type="text"]').type(TODO_ITEM)
    cy.get('button').click()
    // just verify the dispatched action
    cy.get('@dispatch').should('be.calledWith', {
      type: todos.types.ADD,
      payload: { id: 0, text: TODO_ITEM },
    })
  })
})

describe('FilterInfo Component', () => {
  // beforeEach(fixCypressSpec(__filename, window))
  beforeEach(function() {
    // see https://github.com/bahmutov/cypress-react-unit-test/issues/51
    fixReactDomScope(window);
  });

  it('filter completed items', () => {
    // given
    const props = {
      filters: ['All', 'Active', 'Completed'],
    }
    // when
    cy.mountWithRedux(<FilterInfo {...props} />)

    // then
    // cy.get(FilterInfo).toMatchSnapshot()
    cy.contains('Active').click()
    // just verify the dispatched action
    cy.get('@dispatch').should('be.calledWith', {
      type: filter.types.FILTER,
      payload: { status: 'Active' },
    })
  })
})
