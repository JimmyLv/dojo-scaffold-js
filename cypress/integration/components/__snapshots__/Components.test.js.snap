exports[`TodoItem Component > shows an item #0`] = `
<span>test item </span>
`;
